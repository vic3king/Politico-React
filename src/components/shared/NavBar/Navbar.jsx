import React from 'react';
import { Link } from 'react-router-dom';
import '../../../style/navbar.scss';
import PropTypes from 'prop-types';

const NavBar = ({ LiTagOne, LiTagTwo }) => {
  return (
    <React.Fragment>
      <header className="main-header">
        <div>
          <i className="fa fa-balance-scale fa-lg" />
          <Link to="/" className="main-header__brand">
            Politico
          </Link>
        </div>
        <nav className="main-nav">
          <ul className="main-nav__items">
            {LiTagOne}
            {LiTagTwo}
          </ul>
        </nav>
      </header>
    </React.Fragment>
  );
};
NavBar.defaultProps = {
  LiTagOne: (PropTypes.defaultProps = ''),
  LiTagTwo: (PropTypes.defaultProps = ''),
};
NavBar.propTypes = {
  LiTagOne: PropTypes.object,
  LiTagTwo: PropTypes.object,
};

export default NavBar;
