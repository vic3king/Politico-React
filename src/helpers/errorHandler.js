const errorHandler = obj => {
  let error;

  if (typeof obj === 'string') {
    error = obj;
  }

  if (typeof obj === 'object') {
    Object.values(obj).forEach(err => {
      error = err;
    });
  }

  return error;
};

export default errorHandler;
