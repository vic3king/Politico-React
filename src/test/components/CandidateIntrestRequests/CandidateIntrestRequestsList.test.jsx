import React from 'react';
import { shallow } from 'enzyme';
import { notify } from 'react-notify-toast';
import CandidateIntrestRequestsList from '../../../components/CandidateIntrestRequests/CandidateIntrestRequestList';
import CandidateIntrestRequestsCard from '../../../components/CandidateIntrestRequests/CandidateIntrestRequests';
import NavBar from '../../../components/shared/NavBar/Navbar';
import Interest from '../../../services/interest';
import Vote from '../../../services/votes';
import errorHandler from '../../../helpers/errorHandler';

jest.mock('../../../services/interest');
jest.mock('../../../services/votes');
jest.mock('react-notify-toast');

let wrapper;
describe('CandidateIntrestRequestsList component', () => {
  beforeAll(() => {
    Vote.getCandidatesByStatus = jest.fn().mockImplementation(() =>
      Promise.resolve({
        status: 200,
        data: [
          {
            firstname: '',
          },
          {
            firstname: '',
          },
        ],
      })
    );
  });
  beforeEach(() => {
    wrapper = shallow(<CandidateIntrestRequestsList />);
  });
  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should contain necessary components', () => {
    expect(wrapper.find(<NavBar />)).toBeTruthy();
    expect(wrapper.find(<CandidateIntrestRequestsCard />)).toBeTruthy();
    expect(wrapper.find('div')).toBeTruthy();
    expect(wrapper.find('button')).toBeTruthy();
  });

  describe('ComponentDidMount', () => {
    it('should call componentDidMount method', async () => {
      const instance = wrapper.instance();
      await instance.componentDidMount();

      expect(instance.state.candidates).toEqual([
        {
          firstname: '',
        },
        {
          firstname: '',
        },
      ]);
    });
  });

  describe('updateCandidatesState method', () => {
    it('should remove candidate from state', () => {
      const instance = wrapper.instance();
      instance.setState({
        candidates: [
          {
            id: 1,
          },
          {
            id: 2,
          },
        ],
      });

      instance.updateCandidatesState(1);

      expect(
        instance.state.candidates.find(candidate => {
          return candidate.id === 1;
        })
      ).toBeFalsy();
    });
  });

  describe('adminAcceptReject method', () => {
    let instance;
    let event;
    describe('Api call success', () => {
      beforeAll(() => {
        Interest.processRequest = jest.fn().mockImplementation(() =>
          Promise.resolve({
            status: 200,
          })
        );
      });
      beforeEach(() => {
        instance = wrapper.instance();
        event = {
          preventDefault: jest.fn(),
          target: { id: 2 },
        };
      });
      it('should call preventDefault on event', async () => {
        await instance.adminAcceptRequest(event);

        expect(event.preventDefault).toBeCalled();
      });

      it('should call toast the right message on api call success', async () => {
        await instance.adminAcceptRequest(event);
        expect(notify.show).toBeCalledWith('Candidate Approved');
      });
      it('should set loading state to false after successfull api call is made', async () => {
        await instance.adminAcceptRequest(event);
        expect(instance.state.loading).toBeFalsy();
      });
      it('should call the updateCandidatesState method with id from api call', async () => {
        const updateCandidatesState = jest.spyOn(
          instance,
          'updateCandidatesState'
        );
        await instance.adminAcceptRequest(event);
        expect(updateCandidatesState).toBeCalledWith(2);
      });
    });

    describe('Api call 400 errors', () => {
      beforeAll(() => {
        Interest.processRequest = jest.fn().mockImplementation(() =>
          Promise.resolve({
            status: 400,
            errors: {
              message: 'validation error',
            },
          })
        );
      });
      beforeEach(() => {
        instance = wrapper.instance();
        event = {
          preventDefault: jest.fn(),
          target: { id: 2 },
        };
      });

      it('should set loading state to false on validation error', async () => {
        await instance.adminAcceptRequest(event);
        expect(instance.state.loading).toBeFalsy();
      });

      it('should toast the error message', async () => {
        await instance.adminAcceptRequest(event);
        expect(notify.show).toBeCalledWith(
          errorHandler('validation error'),
          'error'
        );
      });
    });
  });

  describe('adminRejectRequest method', () => {
    let instance;
    let event;
    describe('Api call success', () => {
      beforeAll(() => {
        Interest.processRequest = jest.fn().mockImplementation(() =>
          Promise.resolve({
            status: 200,
          })
        );
      });
      beforeEach(() => {
        instance = wrapper.instance();
        event = {
          preventDefault: jest.fn(),
          target: { id: 2 },
        };
      });
      it('should call preventDefault on event', async () => {
        await instance.adminRejectRequest(event);

        expect(event.preventDefault).toBeCalled();
      });

      it('should call toast the right message on api call success', async () => {
        await instance.adminRejectRequest(event);
        expect(notify.show).toBeCalledWith('Candidate Rejected');
      });
      it('should set loading state to false after successfull api call is made', async () => {
        await instance.adminRejectRequest(event);
        expect(instance.state.loading).toBeFalsy();
      });
      it('should call the updateCandidatesState method with id from api call', async () => {
        const updateCandidatesState = jest.spyOn(
          instance,
          'updateCandidatesState'
        );
        await instance.adminRejectRequest(event);
        expect(updateCandidatesState).toBeCalledWith(2);
      });
    });

    describe('Api call 400 errors', () => {
      beforeAll(() => {
        Interest.processRequest = jest.fn().mockImplementation(() =>
          Promise.resolve({
            status: 400,
            errors: {
              message: 'validation error',
            },
          })
        );
      });
      beforeEach(() => {
        instance = wrapper.instance();
        event = {
          preventDefault: jest.fn(),
          target: { id: 2 },
        };
      });

      it('should set loading state to false on validation error', async () => {
        await instance.adminRejectRequest(event);
        expect(instance.state.loading).toBeFalsy();
      });

      it('should toast the error message', async () => {
        await instance.adminRejectRequest(event);
        expect(notify.show).toBeCalledWith(
          errorHandler('validation error'),
          'error'
        );
      });
    });
  });
});
