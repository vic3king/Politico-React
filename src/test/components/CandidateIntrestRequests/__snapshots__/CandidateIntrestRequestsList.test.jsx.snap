// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CandidateIntrestRequestsList component should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CandidateIntrestRequestsList />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <$class
          options={Object {}}
        />,
        false,
        <NavBar
          LiTagOne={
            <LiTag
              to="/logout"
              value="Logout"
            />
          }
          LiTagTwo=""
        />,
        <h4
          className="page-title-results"
        >
          Review Pending Requests
        </h4>,
        <div>
          <div
            className="main_page_title"
          >
            <div
              className="main-grid"
            >
              <CandidateIntrestRequestsCard
                adminAcceptRequest={[Function]}
                adminRejectRequest={[Function]}
                btnClass="btn-accept"
                btnClass1="btn-reject"
                button="Accept"
                fullname=" undefined"
              />
              <CandidateIntrestRequestsCard
                adminAcceptRequest={[Function]}
                adminRejectRequest={[Function]}
                btnClass="btn-accept"
                btnClass1="btn-reject"
                button="Accept"
                fullname=" undefined"
              />
            </div>
          </div>
        </div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "options": Object {},
        },
        "ref": null,
        "rendered": null,
        "type": [MockFunction],
      },
      false,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "LiTagOne": <LiTag
            to="/logout"
            value="Logout"
          />,
          "LiTagTwo": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Review Pending Requests",
          "className": "page-title-results",
        },
        "ref": null,
        "rendered": "Review Pending Requests",
        "type": "h4",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="main_page_title"
          >
            <div
              className="main-grid"
            >
              <CandidateIntrestRequestsCard
                adminAcceptRequest={[Function]}
                adminRejectRequest={[Function]}
                btnClass="btn-accept"
                btnClass1="btn-reject"
                button="Accept"
                fullname=" undefined"
              />
              <CandidateIntrestRequestsCard
                adminAcceptRequest={[Function]}
                adminRejectRequest={[Function]}
                btnClass="btn-accept"
                btnClass1="btn-reject"
                button="Accept"
                fullname=" undefined"
              />
            </div>
          </div>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="main-grid"
            >
              <CandidateIntrestRequestsCard
                adminAcceptRequest={[Function]}
                adminRejectRequest={[Function]}
                btnClass="btn-accept"
                btnClass1="btn-reject"
                button="Accept"
                fullname=" undefined"
              />
              <CandidateIntrestRequestsCard
                adminAcceptRequest={[Function]}
                adminRejectRequest={[Function]}
                btnClass="btn-accept"
                btnClass1="btn-reject"
                button="Accept"
                fullname=" undefined"
              />
            </div>,
            "className": "main_page_title",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <CandidateIntrestRequestsCard
                  adminAcceptRequest={[Function]}
                  adminRejectRequest={[Function]}
                  btnClass="btn-accept"
                  btnClass1="btn-reject"
                  button="Accept"
                  fullname=" undefined"
                />,
                <CandidateIntrestRequestsCard
                  adminAcceptRequest={[Function]}
                  adminRejectRequest={[Function]}
                  btnClass="btn-accept"
                  btnClass1="btn-reject"
                  button="Accept"
                  fullname=" undefined"
                />,
              ],
              "className": "main-grid",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "adminAcceptRequest": [Function],
                  "adminRejectRequest": [Function],
                  "btnClass": "btn-accept",
                  "btnClass1": "btn-reject",
                  "button": "Accept",
                  "fullname": " undefined",
                  "id": undefined,
                  "office": undefined,
                  "party": undefined,
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "adminAcceptRequest": [Function],
                  "adminRejectRequest": [Function],
                  "btnClass": "btn-accept",
                  "btnClass1": "btn-reject",
                  "button": "Accept",
                  "fullname": " undefined",
                  "id": undefined,
                  "office": undefined,
                  "party": undefined,
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": "div",
          },
          "type": "div",
        },
        "type": "div",
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <$class
            options={Object {}}
          />,
          false,
          <NavBar
            LiTagOne={
              <LiTag
                to="/logout"
                value="Logout"
              />
            }
            LiTagTwo=""
          />,
          <h4
            className="page-title-results"
          >
            Review Pending Requests
          </h4>,
          <div>
            <div
              className="main_page_title"
            >
              <div
                className="main-grid"
              >
                <CandidateIntrestRequestsCard
                  adminAcceptRequest={[Function]}
                  adminRejectRequest={[Function]}
                  btnClass="btn-accept"
                  btnClass1="btn-reject"
                  button="Accept"
                  fullname=" undefined"
                />
                <CandidateIntrestRequestsCard
                  adminAcceptRequest={[Function]}
                  adminRejectRequest={[Function]}
                  btnClass="btn-accept"
                  btnClass1="btn-reject"
                  button="Accept"
                  fullname=" undefined"
                />
              </div>
            </div>
          </div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "options": Object {},
          },
          "ref": null,
          "rendered": null,
          "type": [MockFunction],
        },
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "LiTagOne": <LiTag
              to="/logout"
              value="Logout"
            />,
            "LiTagTwo": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Review Pending Requests",
            "className": "page-title-results",
          },
          "ref": null,
          "rendered": "Review Pending Requests",
          "type": "h4",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="main_page_title"
            >
              <div
                className="main-grid"
              >
                <CandidateIntrestRequestsCard
                  adminAcceptRequest={[Function]}
                  adminRejectRequest={[Function]}
                  btnClass="btn-accept"
                  btnClass1="btn-reject"
                  button="Accept"
                  fullname=" undefined"
                />
                <CandidateIntrestRequestsCard
                  adminAcceptRequest={[Function]}
                  adminRejectRequest={[Function]}
                  btnClass="btn-accept"
                  btnClass1="btn-reject"
                  button="Accept"
                  fullname=" undefined"
                />
              </div>
            </div>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <div
                className="main-grid"
              >
                <CandidateIntrestRequestsCard
                  adminAcceptRequest={[Function]}
                  adminRejectRequest={[Function]}
                  btnClass="btn-accept"
                  btnClass1="btn-reject"
                  button="Accept"
                  fullname=" undefined"
                />
                <CandidateIntrestRequestsCard
                  adminAcceptRequest={[Function]}
                  adminRejectRequest={[Function]}
                  btnClass="btn-accept"
                  btnClass1="btn-reject"
                  button="Accept"
                  fullname=" undefined"
                />
              </div>,
              "className": "main_page_title",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <CandidateIntrestRequestsCard
                    adminAcceptRequest={[Function]}
                    adminRejectRequest={[Function]}
                    btnClass="btn-accept"
                    btnClass1="btn-reject"
                    button="Accept"
                    fullname=" undefined"
                  />,
                  <CandidateIntrestRequestsCard
                    adminAcceptRequest={[Function]}
                    adminRejectRequest={[Function]}
                    btnClass="btn-accept"
                    btnClass1="btn-reject"
                    button="Accept"
                    fullname=" undefined"
                  />,
                ],
                "className": "main-grid",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "adminAcceptRequest": [Function],
                    "adminRejectRequest": [Function],
                    "btnClass": "btn-accept",
                    "btnClass1": "btn-reject",
                    "button": "Accept",
                    "fullname": " undefined",
                    "id": undefined,
                    "office": undefined,
                    "party": undefined,
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "adminAcceptRequest": [Function],
                    "adminRejectRequest": [Function],
                    "btnClass": "btn-accept",
                    "btnClass1": "btn-reject",
                    "button": "Accept",
                    "fullname": " undefined",
                    "id": undefined,
                    "office": undefined,
                    "party": undefined,
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
              ],
              "type": "div",
            },
            "type": "div",
          },
          "type": "div",
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
